### PR TITLE
Fix `chainerx::Dot` edge cases with empty arrays

### PR DESCRIPTION
**Pull request by hvy**
Fixed `chainerx::Dot` for empty arrays and simplified implementation. 

This partially overlaps with https://github.com/chainer/chainer/pull/8007.

### CODE REVIEW

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 1]
```diff
@@ -42,18 +42,10 @@ Array Dot(const Array& a, const Array& b, absl::optional<Dtype> out_dtype) {
...
-        std::copy(axes_index.begin(), axes_index.end() - 2, std::back_inserter(b_transpose_axes));
-        std::reverse_copy(axes_index.end() - 2, axes_index.end(), std::back_inserter(b_transpose_axes));
-
-        Axes axes(b_transpose_axes.begin(), b_transpose_axes.end());
-        modified_b = b.Transpose(axes);
+        modified_b = Swapaxes(b, -1, -2);
         std::copy(modified_b.shape().begin(), modified_b.shape().end() - 1, std::back_inserter(out_shape));
 
-        modified_b = modified_b.Reshape({-1, modified_b.shape().back()});
+        modified_b = modified_b.Reshape({b.GetTotalSize() > 0 ? -1 : 0, modified_b.shape().back()});
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: We cannot reshape with `-1` if the array is empty https://github.com/chainer/chainer/pull/8018.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by hvy** [comment]
Should we backport this PR?

**Comment by niboshi** [comment]
Ditto: https://github.com/chainer/chainer/pull/8018#issuecomment-524226638

**Pull request merged**